File: jiuzhang/Nine Chapters/4 Dynamic Programming I/java/WordBreak_003.java
// To fix: cand & newcand [from Set -> ArrayList]

public class Solution {
    /**
     * @param s: A string s
     * @param dict: A dictionary of words dict
     */
    public boolean wordBreak(String s, Set<String> dict) {
        // write your code here   
        // write your code here   
        if (s.equals("") && dict.isEmpty()) {
            return true;
        }

        HashMap<Integer, Set<String>> lendict = new HashMap<Integer, Set<String>>();
        ArrayList<Integer> lenlist = new ArrayList<Integer>();

        for (String word : dict) {
            int l = word.length();
            if (lendict.containsKey(l)) {
                lendict.get(l).add(word);
            } else {
                HashSet<String> tmp = new HashSet<String>();
	        	tmp.add(word);
	        	lendict.put(l, tmp);
                lenlist.add(l);
            }
        }

        Collections.sort(lenlist);
        HashSet<Integer> cand = new HashSet<Integer>();
        for (int l : lenlist) {
            if (l <= s.length()) {
                String substr = s.substring(0, l);
                if (lendict.get(l).contains(substr)) {
                    if (l == s.length()) {
                        return true;
                    }
                    cand.add(l - 1);
                }
            } else {
                break;
            }
        }
        
        boolean[] isVisited = new boolean[s.length()];
    
        while (! cand.isEmpty()) {
            HashSet<Integer> newcand = new HashSet<Integer>();
            
            for (int c : cand) {
                for (int l : lenlist) {
                    int idx = c + l;
                    if (idx > s.length() - 1) {
                        break;
                    } else {
                        String word = s.substring(c + 1, idx + 1);
                        if ((! isVisited[idx]) && (lendict.get(l).contains(word))) {
                            newcand.add(idx);
                            isVisited[idx] = true;
                            if (idx == s.length() - 1) {
                                return true;
                            }
                        }
                        
                    }
                }
            }
            cand = newcand;
        }

        return false;
    }
}

File: oj/jiuzhang/Nine Chapters/4 Dynamic Programming I/py/WordBreak_003.py
# Complicated, not clear and straightforward
# Have to handle special cases
# O(n^2), same speed on leetcode tests as 002
# {Leetcode: AC, Lintcode : AC}

class Solution:
    # @param s: A string s
    # @param dict: A dictionary of words dict
    def wordBreak(self, s, dict):
        # write your code here
        if s == '' and dict == {}:
            return True
        lendict, lenlist = {}, []
        for word in dict:
            l = len(word)
            if l in lendict:
                lendict[l].add(word)
            else:
                lendict[l] = set([word])
                lenlist.append(l)
        lenlist.sort()
        cand = []
        for l in lenlist:
            if l <= len(s):
                if s[:l] in lendict[l]:
                    if l == len(s):
                        return True
                    cand.append(l - 1)
            else:
                break

        notVisited = [True] * len(s)

        while len(cand) != 0:
            newcand = []
            for c in cand:
                for l in lenlist:
                    idx = c + l
                    if idx > len(s) - 1:
                        break
                    else:
                        word = s[c + 1: idx + 1]
                        if notVisited[idx] and word in lendict[l]:
                            newcand.append(idx)
                            notVisited[idx] = False
                            if idx == len(s) - 1:
                                return True
            cand = newcand

        return False
